README.md
# egg-logger

[![NPM version][npm-image]][npm-url]
[![build status][github-action-image]][github-action-url]
[![Test coverage][codecov-image]][codecov-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/egg-logger.svg?style=flat-square
[npm-url]: https://npmjs.org/package/egg-logger
[github-action-image]: https://github.com/eggjs/egg-logger/workflows/Node.js%20CI/badge.svg
[github-action-url]: https://github.com/eggjs/egg-logger/actions
[codecov-image]: https://codecov.io/github/eggjs/egg-logger/coverage.svg?branch=master
[codecov-url]: https://codecov.io/github/eggjs/egg-logger?branch=master
[snyk-image]: https://snyk.io/test/npm/egg-logger/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/egg-logger
[download-image]: https://img.shields.io/npm/dm/egg-logger.svg?style=flat-square
[download-url]: https://npmjs.org/package/egg-logger

Egg logger.

![diagram](diagram.png)

Including two base class, `Logger` and `Transport`:

- Transport: Save log to file, stdout/stderr and network.
- Logger: A logger can contains multi transports.

---

## Install

```bash
$ npm i egg-logger
```

## Usage

Create a `Logger` and add a file `Transport`.

```js
const Logger = require('egg-logger').Logger;
const FileTransport = require('egg-logger').FileTransport;
const ConsoleTransport = require('egg-logger').ConsoleTransport;

const logger = new Logger();
logger.set('file', new FileTransport({
  file: '/path/to/file',
  level: 'INFO',
}));
logger.set('console', new ConsoleTransport({
  level: 'DEBUG',
}));
logger.debug('debug foo'); // only output to stdout
logger.info('info foo');
logger.warn('warn foo');
logger.error(new Error('error foo'));
```

### Enable / Disable Transport

```js
logger.disable('file');
logger.info('info'); // output nothing
logger.enable('file');
logger.info('info'); // output 'info' string
```

### Duplicate

Duplicate error log to other logger.

Accept an `options.excludes` to special whether excludes some tranports.

```js
logger.duplicate('error', errorLogger, { excludes: [ 'console' ]});
logger.error(new Error('print to errorLogger')); // will additional call `errorLogger.error`
```

### Redirect

Redirect special level log to other logger.

```js
oneLogger.redirect('debug', debugLogger); // all debug level logs of `oneLogger` will delegate to debugLogger
```

### Reload

```js
logger.reload(); // will close the exists write stream and create a new one.
```

### Custom Transport

You can make your own `Transport` for logging，e.g.: send log to your logging server.

```js
const urllib = require('urllib');
const Transport = require('egg-logger').Transport;

class UrllibTransport extends Transport {

  log(level, args, meta) {
    const msg = super.log(level, args, meta);
    return urllib.request('url?msg=' + msg);
  }
}

const logger = new Logger();
logger.set('remote', new UrllibTransport({
  level: 'DEBUG',
}));
logger.info('info');
```

## Console logger level

set environment NODE_CONSOLE_LOGGRE_LEVEL = 'INFO' | 'WARN' | 'ERROR'


## License

[MIT](LICENSE)
<!-- GITCONTRIBUTOR_START -->

## Contributors

|[<img src="https://avatars.githubusercontent.com/u/156269?v=4" width="100px;"/><br/><sub><b>fengmk2</b></sub>](https://github.com/fengmk2)<br/>|[<img src="https://avatars.githubusercontent.com/u/985607?v=4" width="100px;"/><br/><sub><b>dead-horse</b></sub>](https://github.com/dead-horse)<br/>|[<img src="https://avatars.githubusercontent.com/u/360661?v=4" width="100px;"/><br/><sub><b>popomore</b></sub>](https://github.com/popomore)<br/>|[<img src="https://avatars.githubusercontent.com/u/227713?v=4" width="100px;"/><br/><sub><b>atian25</b></sub>](https://github.com/atian25)<br/>|[<img src="https://avatars.githubusercontent.com/u/2160731?v=4" width="100px;"/><br/><sub><b>mansonchor</b></sub>](https://github.com/mansonchor)<br/>|[<img src="https://avatars.githubusercontent.com/u/5856440?v=4" width="100px;"/><br/><sub><b>whxaxes</b></sub>](https://github.com/whxaxes)<br/>|
| :---: | :---: | :---: | :---: | :---: | :---: |
|[<img src="https://avatars.githubusercontent.com/u/3367820?v=4" width="100px;"/><br/><sub><b>Jeff-Tian</b></sub>](https://github.com/Jeff-Tian)<br/>|[<img src="https://avatars.githubusercontent.com/u/19274678?v=4" width="100px;"/><br/><sub><b>andy-ms</b></sub>](https://github.com/andy-ms)<br/>|[<img src="https://avatars.githubusercontent.com/u/1474688?v=4" width="100px;"/><br/><sub><b>luckydrq</b></sub>](https://github.com/luckydrq)<br/>|[<img src="https://avatars.githubusercontent.com/u/2842176?v=4" width="100px;"/><br/><sub><b>XadillaX</b></sub>](https://github.com/XadillaX)<br/>|[<img src="https://avatars.githubusercontent.com/u/8201516?v=4" width="100px;"/><br/><sub><b>linrf</b></sub>](https://github.com/linrf)<br/>|[<img src="https://avatars.githubusercontent.com/u/30541930?v=4" width="100px;"/><br/><sub><b>duqingyu</b></sub>](https://github.com/duqingyu)<br/>|
[<img src="https://avatars.githubusercontent.com/u/7779883?v=4" width="100px;"/><br/><sub><b>lix059</b></sub>](https://github.com/lix059)<br/>|[<img src="https://avatars.githubusercontent.com/u/2675419?v=4" width="100px;"/><br/><sub><b>congyuandong</b></sub>](https://github.com/congyuandong)<br/>|[<img src="https://avatars.githubusercontent.com/u/1763067?v=4" width="100px;"/><br/><sub><b>waitingsong</b></sub>](https://github.com/waitingsong)<br/>

This project follows the git-contributor [spec](https://github.com/xudafeng/git-contributor), auto updated at `Sat Dec 10 2022 21:18:13 GMT+0800`.

<!-- GITCONTRIBUTOR_END -->

History.md

3.0.1 / 2022-12-10
==================

**fixes**
  * [[`1bd4715`](http://github.com/eggjs/egg-logger/commit/1bd4715d617874373f3d750f4de7e0d8fb3b2653)] - 🐛 FIX: Add localStorage options to EggLoggers (#81) (fengmk2 <<fengmk2@gmail.com>>)

**others**
  * [[`804ff20`](http://github.com/eggjs/egg-logger/commit/804ff2024f92c6f4da2182a4afec4b506da2f805)] - 📖 DOC: Update contributors (fengmk2 <<fengmk2@gmail.com>>)

3.0.0 / 2022-12-10
==================

**features**
  * [[`0d4c55d`](http://github.com/eggjs/egg-logger/commit/0d4c55d37e1569950856832b56dbc187df677bf8)] - 📦 NEW: [BREAKING] support asyncLocalStorage to get ctx (#80) (fengmk2 <<fengmk2@gmail.com>>)

**others**
  * [[`5776bad`](http://github.com/eggjs/egg-logger/commit/5776bad40f57d04fb73dd852eb69b6b3947c8a8a)] - Create codeql.yml (fengmk2 <<fengmk2@gmail.com>>)

2.9.1 / 2022-11-15
==================

**fixes**
  * [[`ec97edd`](http://github.com/eggjs/egg-logger/commit/ec97edd6d165f559f8c86bf1573cd183b156fd06)] - 🐛 FIX: Ignore error log after stream close on unittest env (#79) (fengmk2 <<fengmk2@gmail.com>>)

2.9.0 / 2022-07-27
==================

**features**
  * [[`493a595`](http://github.com/eggjs/egg-logger/commit/493a5952a4a7c5314a457099a9e9adfd1d1525ce)] - feat: support custom concentrateLoggerName options (#78) (mansonchor.github.com <<mansonchor1987@gmail.com>>)

2.8.0 / 2022-03-22
==================

**features**
  * [[`6f7dcbb`](http://github.com/eggjs/egg-logger/commit/6f7dcbb1ec92b0242c5adc43a41681c3848d4c62)] - feat: Only output logs in JSON file format is supported (#77) (duqingyu <<1065161421@qq.com>>)

2.7.1 / 2021-12-30
==================

**fixes**
  * [[`fe93575`](http://github.com/eggjs/egg-logger/commit/fe935752579ddbf60c6125999faa250ba36e3c35)] - fix: EggLoggers constructor typing (#75) (Lockonlin <<531244799@qq.com>>)

2.7.0 / 2021-12-05
==================

**features**
  * [[`f42cb95`](http://github.com/eggjs/egg-logger/commit/f42cb950ab7e00872c464cff87e7b2b77da5c391)] - feat: support performance timer for more precise milliseconds (#74) (fengmk2 <<fengmk2@gmail.com>>)

2.6.2 / 2021-08-02
==================

**fixes**
  * [[`62d724f`](http://github.com/eggjs/egg-logger/commit/62d724fc2b4b2a22c2d554dbfc4c882b766ed630)] - fix: remove defineProperty to improve performance (#72) (Yiyu He <<dead_horse@qq.com>>)

2.6.1 / 2021-04-07
==================

**fixes**
  * [[`0d3d4e4`](http://github.com/eggjs/egg-logger/commit/0d3d4e461d802340ec989897694c9f9af07b4f3e)] - fix: use FrameworkBaseError.isFrameworkError to judge (#71) (mansonchor.github.com <<mansonchor1987@gmail.com>>)

2.6.0 / 2021-03-23
==================

**features**
  * [[`d414828`](http://github.com/eggjs/egg-logger/commit/d414828d6e511c2be6be7586380ecb917281c238)] - feat: logger formatError should support FrameworkErrorFormater (#70) (mansonchor.github.com <<mansonchor@126.com>>)

**others**
  * [[`33e1e74`](http://github.com/eggjs/egg-logger/commit/33e1e743d932a1897a2588ba5d4c64f05ce9d437)] - refactor: new Buffer() API refactor (#69) (mansonchor.github.com <<mansonchor@126.com>>)
  * [[`5303d31`](http://github.com/eggjs/egg-logger/commit/5303d31bc1512e85561e000836143bc7daae20d9)] - test: only run tests on github actions (#68) (fengmk2 <<fengmk2@gmail.com>>)

2.5.0 / 2021-02-11
==================

**features**
  * [[`81e2328`](http://github.com/eggjs/egg-logger/commit/81e2328a76fc26a4f9a1df0915127ad9e17600ec)] - feat: support console logger default level on env (#67) (lx <<lix059@gmail.com>>)

**others**
  * [[`f987c06`](http://github.com/eggjs/egg-logger/commit/f987c0621b6b272ba9d56caa8af59b65c1251930)] - chore: github action (#63) (TZ | 天猪 <<atian25@qq.com>>)

2.4.2 / 2020-04-14
==================

**others**
  * [[`6fbe6c3`](http://github.com/eggjs/egg-logger/commit/6fbe6c34b7456c683342a51e360f667128afe315)] - chore(typings): add concentrateError to EggLoggerOptions (#62) (scott <<congyuandong@gmail.com>>)
  * [[`a1b6c85`](http://github.com/eggjs/egg-logger/commit/a1b6c85ad13f7fa39dc3fc72029c39604b05cd69)] - chore: update travis (#60) (TZ | 天猪 <<atian25@qq.com>>)

2.4.1 / 2019-03-19
==================

**fixes**
  * [[`d048015`](http://github.com/eggjs/egg-logger/commit/d048015f63dde2917cb51f8f30b87a320b2d892b)] - fix: duplicate should ignore some transports (#45) (TZ | 天猪 <<atian25@qq.com>>)

2.4.0 / 2019-03-14
==================

**features**
  * [[`0cbbe47`](http://github.com/eggjs/egg-logger/commit/0cbbe47efb10897ac55a892bc670d3924a1ddfb5)] - feat: support contextFormatter (#51) (Haoliang Gao <<sakura9515@gmail.com>>)

2.3.2 / 2019-02-28
==================

**fixes**
  * [[`f51216b`](http://github.com/eggjs/egg-logger/commit/f51216bd84d48f75e44ebe2e1208c308d1969dce)] - fix: crash when format a getter-only error stack (#50) (Khaidi Chu <<i@2333.moe>>)

**others**
  * [[`858a8b4`](http://github.com/eggjs/egg-logger/commit/858a8b46495fed21efb4215e36764b3030e573ee)] - chore: comments typo fix (#47) (Jeff <<jeff.tian@outlook.com>>)

2.3.1 / 2019-01-03
==================

  * fix: should not duplicate to console (#43)
  * chore: modify engine version (#42)

2.3.0 / 2019-01-01
==================

**features**
  * [[`22cf355`](http://github.com/eggjs/egg-logger/commit/22cf355d7ae976c8502745a8e8d73748af885637)] - feat: file path support relative path (#41) (TZ | 天猪 <<atian25@qq.com>>)

**others**
  * [[`0262ac2`](http://github.com/eggjs/egg-logger/commit/0262ac204826f1a582bd63273a43cfe86c723ad3)] - chore: fix typo (#40) (Jeff <<jeff.tian@outlook.com>>)

2.2.1 / 2018-12-26
==================

  * fix: logger should extend Map (#39)

2.2.0 / 2018-12-17
==================

**features**
  * [[`d578857`](http://github.com/eggjs/egg-logger/commit/d57885737c4311f17a76713abcd5459523ff92a4)] - feat: support concentrateError and set default to duplicate (#38) (TZ | 天猪 <<atian25@qq.com>>)

2.1.0 / 2018-12-07
==================

**features**

benchmark/index.js
'use strict';

const { performance } = require('perf_hooks');
const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const path = require('path');
const Logger = require('..').EggLogger;
const ContextLogger = require('..').EggContextLogger;

const loggerWithJSON = new Logger({
  file: path.join(__dirname, '../.logs', 'withjson.log'),
  jsonFile: path.join(__dirname, '../.logs', 'withjson.json.log'),
  consoleLevel: 'NONE',
  encoding: 'utf8',
});

const loggerWithoutJSON = new Logger({
  file: path.join(__dirname, '../.logs', 'withoutjson.log'),
  encoding: 'utf8',
  consoleLevel: 'NONE',
});

const ctx = {
  userId: '2088102117755972',
  logonId: 'money02@alitest.com',
  tracer: {
    traceId: 'c0a8016714537221365031005',
  },
  starttime: Date.now(),
  ip: '127.0.0.1',
  method: 'GET',
  url: '/',
};
const contextLogger = new ContextLogger(ctx, loggerWithoutJSON);
const contextLoggerWithJSON = new ContextLogger(ctx, loggerWithJSON);

const ctxUsingPerformance = {
  userId: '2088102117755972',
  logonId: 'money02@alitest.com',
  tracer: {
    traceId: 'c0a8016714537221365031005',
  },
  performanceStarttime: performance.now(),
  ip: '127.0.0.1',
  method: 'GET',
  url: '/',
};
const contextLoggerUsingPerformance = new ContextLogger(ctxUsingPerformance, loggerWithoutJSON);
const contextLoggerWithJSONUsingPerformance = new ContextLogger(ctxUsingPerformance, loggerWithJSON);

const err = new Error('test error');
err.data = { foo: 'bar' };
err.code = 'MOCK';

const suite = new Benchmark.Suite();

suite
  .add('logger.error(err)', () => {
    loggerWithoutJSON.error(err);
  })
  .add('logger.info(message)', () => {
    loggerWithoutJSON.info('hello, %s', 'world');
  })
  .add('logger.write(message)', () => {
    loggerWithoutJSON.write('hello, world');
  })
  .add('contextLogger.info(message)', () => {
    contextLogger.info('hello, %s', 'world');
  })
  .add('performanceStarttime: contextLogger.info(message)', () => {
    contextLoggerUsingPerformance.info('hello, %s', 'world');
  })
  .add('logger.error(err) with JSON', () => {
    loggerWithJSON.error(err);
  })
  .add('logger.info(message) with JSON', () => {
    loggerWithJSON.info('hello, %s', 'world');
  })
  .add('logger.write(message) with JSON', () => {
    loggerWithJSON.write('hello, world');
  })
  .add('contextLogger.info(message) with JSON', () => {
    contextLoggerWithJSON.info('hello, %s', 'world');
  })
  .add('performanceStarttime: contextLogger.info(message) with JSON', () => {
    contextLoggerWithJSONUsingPerformance.info('hello, %s', 'world');
  })
  .on('cycle', event => {
    benchmarks.add(event.target);
  })
  .on('start', () => {
    loggerWithJSON.disable('console');
    loggerWithoutJSON.disable('console');
    console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
  })
  .on('complete', () => {
    benchmarks.log();
    process.exit(0);
  })
  .run({ async: false });

// node version: v16.13.0, date: Sat Nov 20 2021 14:05:24 GMT+0800 (中国标准时间)
// Starting...
// 10 tests completed.

// logger.error(err)                                           x   378,794 ops/sec ±1.35% (91 runs sampled)
// logger.info(message)                                        x 2,026,394 ops/sec ±1.34% (97 runs sampled)
// logger.write(message)                                       x 7,454,818 ops/sec ±4.36% (86 runs sampled)
// contextLogger.info(message)                                 x 1,266,091 ops/sec ±3.83% (88 runs sampled)
// performanceStarttime: contextLogger.info(message)           x 1,081,355 ops/sec ±3.03% (86 runs sampled)
// logger.error(err) with JSON                                 x   121,808 ops/sec ±2.62% (90 runs sampled)
// logger.info(message) with JSON                              x   513,678 ops/sec ±2.08% (91 runs sampled)
// logger.write(message) with JSON                             x 4,120,955 ops/sec ±2.62% (93 runs sampled)
// contextLogger.info(message) with JSON                       x   420,963 ops/sec ±3.39% (88 runs sampled)
// performanceStarttime: contextLogger.info(message) with JSON x   382,180 ops/sec ±3.46% (88 runs sampled)

// node version: v14.18.1, date: Sat Nov 20 2021 14:10:56 GMT+0800 (中国标准时间)
// Starting...
// 10 tests completed.

// logger.error(err)                                           x   283,941 ops/sec ±0.93% (95 runs sampled)
// logger.info(message)                                        x 1,204,237 ops/sec ±0.58% (92 runs sampled)
// logger.write(message)                                       x 5,056,509 ops/sec ±0.69% (92 runs sampled)
// contextLogger.info(message)                                 x   886,979 ops/sec ±1.05% (93 runs sampled)
// performanceStarttime: contextLogger.info(message)           x   635,255 ops/sec ±1.34% (92 runs sampled)
// logger.error(err) with JSON                                 x    93,043 ops/sec ±1.12% (91 runs sampled)
// logger.info(message) with JSON                              x   329,611 ops/sec ±0.88% (92 runs sampled)
// logger.write(message) with JSON                             x 2,483,464 ops/sec ±3.14% (87 runs sampled)
// contextLogger.info(message) with JSON                       x   308,370 ops/sec ±3.17% (83 runs sampled)
// performanceStarttime: contextLogger.info(message) with JSON x   282,286 ops/sec ±1.34% (93 runs sampled)

// node version: v12.22.7, date: Sat Nov 20 2021 14:09:42 GMT+0800 (GMT+08:00)
// Starting...
// 10 tests completed.

// logger.error(err)                                           x   291,964 ops/sec ±0.81% (94 runs sampled)
// logger.info(message)                                        x 1,314,201 ops/sec ±0.93% (92 runs sampled)
// logger.write(message)                                       x 5,628,384 ops/sec ±0.79% (94 runs sampled)
// contextLogger.info(message)                                 x   900,788 ops/sec ±1.42% (89 runs sampled)
// performanceStarttime: contextLogger.info(message)           x   669,698 ops/sec ±1.50% (89 runs sampled)
// logger.error(err) with JSON                                 x    97,022 ops/sec ±1.57% (88 runs sampled)
// logger.info(message) with JSON                              x   357,988 ops/sec ±1.03% (91 runs sampled)
// logger.write(message) with JSON                             x 2,917,091 ops/sec ±1.10% (87 runs sampled)
// contextLogger.info(message) with JSON                       x   341,035 ops/sec ±1.62% (91 runs sampled)
// performanceStarttime: contextLogger.info(message) with JSON x   295,930 ops/sec ±1.42% (90 runs sampled)

// node version: v10.24.1, date: Sat Nov 20 2021 14:08:19 GMT+0800 (GMT+08:00)
// Starting...
// 10 tests completed.

// logger.error(err)                                           x   184,461 ops/sec ±5.58% (82 runs sampled)
// logger.info(message)                                        x 1,279,248 ops/sec ±2.01% (83 runs sampled)
// logger.write(message)                                       x 5,638,776 ops/sec ±1.52% (80 runs sampled)
// contextLogger.info(message)                                 x   845,329 ops/sec ±1.60% (81 runs sampled)
// performanceStarttime: contextLogger.info(message)           x   662,304 ops/sec ±2.72% (82 runs sampled)
// logger.error(err) with JSON                                 x    67,005 ops/sec ±2.26% (81 runs sampled)
// logger.info(message) with JSON                              x   344,886 ops/sec ±1.85% (89 runs sampled)
// logger.write(message) with JSON                             x 2,864,012 ops/sec ±1.66% (86 runs sampled)
// contextLogger.info(message) with JSON                       x   254,776 ops/sec ±1.71% (89 runs sampled)
// performanceStarttime: contextLogger.info(message) with JSON x   229,181 ops/sec ±1.62% (89 runs sampled)

// node version: v8.17.0, date: Sat Nov 20 2021 14:06:59 GMT+0800 (CST)
// Starting...
// 10 tests completed.

// logger.error(err)                                           x   173,509 ops/sec ±2.36% (78 runs sampled)
// logger.info(message)                                        x 1,418,705 ops/sec ±2.04% (90 runs sampled)
// logger.write(message)                                       x 4,905,138 ops/sec ±1.46% (93 runs sampled)
// contextLogger.info(message)                                 x   945,955 ops/sec ±3.21% (85 runs sampled)
// performanceStarttime: contextLogger.info(message)           x   740,081 ops/sec ±3.33% (83 runs sampled)
// logger.error(err) with JSON                                 x    68,264 ops/sec ±2.70% (81 runs sampled)
// logger.info(message) with JSON                              x   371,275 ops/sec ±2.79% (90 runs sampled)
// logger.write(message) with JSON                             x 2,497,323 ops/sec ±5.17% (81 runs sampled)
// contextLogger.info(message) with JSON                       x   260,842 ops/sec ±3.12% (85 runs sampled)
// performanceStarttime: contextLogger.info(message) with JSON x   226,953 ops/sec ±3.58% (82 runs sampled)

// node version: v4.4.7, date: Tue Jul 26 2016 00:27:29 GMT+0800 (CST)
// Starting...
// 8 tests completed.
//
// logger.error(err)                     x  29,936 ops/sec ±4.01% (75 runs sampled)
// logger.info(message)                  x 163,694 ops/sec ±4.10% (76 runs sampled)
// logger.write(message)                 x 390,052 ops/sec ±4.74% (76 runs sampled)
// contextLogger.info(message)           x 136,986 ops/sec ±4.74% (78 runs sampled)
// logger.error(err) with JSON           x  13,327 ops/sec ±2.43% (75 runs sampled)
// logger.info(message) with JSON        x  61,547 ops/sec ±11.50% (70 runs sampled)
// logger.write(message) with JSON       x 180,179 ops/sec ±8.81% (67 runs sampled)
// contextLogger.info(message) with JSON x  54,134 ops/sec ±3.59% (76 runs sampled)
